### ISSUE TITLE
Pytorch Mobile - Drastically different output for the same model between PC and Android

### ISSUE THREAD

**Issue by yonatankarimish** [pytorch/pytorch#53650]
## 🐛 Bug

I just finished training a model with good results, and serialized it using torch.jit.trace(). However, the model performs very poorly when deployed to my Android phone.

While investigating the cause of the performance difference between devices, I ran both models (pc and android) using a zero-tensor as input. To my surprise, I found out the model outputs are totally different.

This would explain the poor results (since the mobile model essentially outputs nonsense values), which I hope you can take a look at.

## To Reproduce

1. Download the attached zip file and extract the model from inside (essentially an encoder from an encoder-decoder pair): [vox_encoder.zip](https://github.com/pytorch/pytorch/files/6111352/vox_encoder.zip)
2. On a PC running Python: run the following code:

```
imported_encoder = torch.jit.load("path/to/the/model/you/just/unzipped")
imported_encoder.forward(torch.zeros(1, 1, 138, 80, dtype=torch.float))
```

3. On an Android device running Java, run the following code:

```
Module voxEncoder = Module.load(assetFilePath(this, "path/to/the/model/you/just/unzipped"));
long[] tensorShape = new long[]{1, 1, 138, 80};
IValue spectroTensor = IValue.from(Tensor.fromBlob(new float[(int)(138*80)], tensorShape));
IValue[] encoderResults = voxEncoder.forward(spectroTensor).toTuple();
IValue encoderOutput = encoderResults[0];
```

(The `assetFilePath()` method is a copy-and-paste of a method from the [official demo app](https://github.com/pytorch/android-demo-app/blob/79c4a74aad2045b8dc16947c7b7f85490fa1cfef/NativeApp/app/src/main/java/org/pytorch/nativeapp/MainActivity.java))

4. The values returned by both environments are different, even though we just ran the same model twice.

This is the output from the Python environment (Deployed on a PC)
![image](https://user-images.githubusercontent.com/24466117/110537743-c09ca400-812b-11eb-9433-6a62a1e1195e.png)

This is the output from the Java environment (Deployed an an Android phone)
![image](https://user-images.githubusercontent.com/24466117/110537864-eb86f800-812b-11eb-8e47-21d3f026a2bb.png)

## Expected behavior

The model should output the same values, given the same inputs, across environments. Otherwise, no amount of training will make it of any use.

## Environment

PC environment: 
 - PyTorch Version (e.g., 1.0): 1.8.0
 - OS (e.g., Linux): Linux Ubuntu 16.04
 - How you installed PyTorch (`conda`, `pip`, source): pip (Google collaboratory automatically provided)
 - Python version: 3.6 (Bundled with Google Colaboratory)
 - CUDA/cuDNN version: 11.2
 - GPU models and configuration: Tesla V100-SXM2

Android environment
 - Java 8
 - using org.pytorch:pytorch_android:1.8.0

## Additional context
The code for the model is as follows:
```
class Encoder(torch.nn.Module):
    def __init__(self):
        super(Encoder, self).__init__()

        self.conv_1 = nn.Conv2d(in_channels=1, out_channels=10, kernel_size=(5,1), padding=(2,0))
        self.relu_1 = nn.LeakyReLU()
        self.bn_1 = nn.BatchNorm2d(num_features=10) 

        self.conv_2 = nn.Conv2d(in_channels=10, out_channels=1, kernel_size=(5,1), padding=(2,0))
        self.relu_2 = nn.LeakyReLU()
        self.bn_2 = nn.BatchNorm2d(num_features=1)

        self.encoder = nn.GRU(input_size=80, hidden_size=64, batch_first=True, bidirectional=True, num_layers=2)

    def forward(self, spectrogram):
        #Conv block 1
        x = spectrogram
        x = self.conv_1(x)
        x = self.relu_1(x)
        x = self.bn_1(x)

        #Conv block 2
        x = self.conv_2(x)
        x = self.relu_2(x)
        x = self.bn_2(x)

        #Encoder
        x = torch.squeeze(x, dim=1)
        encoder_output, enc_h = self.encoder(x)

        #Reshape hidden state to match the dimensions expected by the decoder
        h_layers_twice, h_batch_size, h_output_dim = enc_h.shape
        unidirectional_h = enc_h.reshape((h_layers_twice // 2, h_batch_size, h_output_dim * 2))

        return encoder_output, unidirectional_h
```

The model was exported (then zipped to provide the attached file) using the following code:
```
encoder = encoder.eval()
mock_encoder_input = torch.tensor(np.random.rand(1, 1, 138, 80), dtype=torch.float)
traced_encoder = torch.jit.trace(encoder, mock_encoder_input)
traced_encoder.save("path/to/storage")
```

I hope the code snippets + the model I provided are enough. However, I do realize they might be a tad too little, so if any more code / files are required I will try and provide them upon request.

**Comment by yonatankarimish** [comment]
Well figured it out, and it [had nothing to do with PyTorch](https://stackoverflow.com/questions/5762547/why-arent-my-asset-files-updated)…

In short - Android will not update asset files when these are changed. You must manually clear the caches through the phone/emulator settings => apps menu (sigh…)

The app was running an older model (one of many trials), causing the discrepancies I saw.

Feel free to close the issue, and sorry for the trouble.

**Comment by rashmi-learning** [comment]
I am experiencing a drastic change in accuracy while running the model on an android device. Can anyone help with how can I improve the current torchscript model accuracy? On my PC I am getting an accuracy of 99% for the trained yolov5 model. Or Is there any such information available on how much accuracy drops when we convert the torch model to torchscript? Thank you

**Comment by yonatankarimish** [comment]
@rashmi-learning Model accuracy should not drop just because you changed the environment.

A few "simple-but-might-just-work" steps worth trying:
1) Try uninstalling and reinstalling the app
2) Try clearing the app caches using the method I described in my comment above
3) Try passing the same image to each of the two models (pc and mobile), and verify it outputs the same model
4) Try initializing a new model in the pc (no training) and trace it into torchscript. Check if the outputs change in the mobile environment after you replace the model.

**Comment by rashmi-learning** [comment]
@yonatankarimish Thanks for your response. How can we check the converted torchscript model on PC as yolov5 does not support torchscript format? I am new to this. Any help will be appreciated. Thank you

**Comment by yonatankarimish** [comment]
@rashmi-learning What I previously unserstood is that you trained a yolov5 model using pytorch on your pc, then converted it to torchscript using torch.jit.trace().

However, now you claim the model you trained cannot be converted to torchscript.

In that case, how do you place the trained model in a mobile environment?

**Comment by rashmi-learning** [comment]
@yonatankarimish Yes I trained the yolov5 model on the pc and then converted it to torchscript using torch.jit.trace(). Here is the code for your reference:
        jit_fn = opt.weights.replace('.pt', '.torchscript.pt')
        lite_fn = opt.weights.replace('.pt', '.torchscript.ptl')

        ts = torch.jit.trace(model, img)
        from torch.utils.mobile_optimizer import optimize_for_mobile
        ts = optimize_for_mobile(ts)
        ts.save(jit_fn)
        #ts.save(lite_fn)
        ts._save_for_lite_interpreter(lite_fn)

For the android app, i am using torchscript.ptl file. I have checked with similar images on PC as well as on Android. The result has too much diffrence.

**Comment by rashmi-learning** [comment]
@yonatankarimish Taking help of this repo https://github.com/pytorch/android-demo-app/tree/master/ObjectDetection for building the application and converting the yolov5 weights to torchscript.ptl format.

**Comment by yonatankarimish** [comment]
@rashmi-learning Please try all four suggestions I made earlier. It would be much easier to help you if you give all the results

**Comment by rashmi-learning** [comment]
@yonatankarimish Thanks for the suggestions.

**Comment by ArtanBerisha1** [comment]
I had a similar problem, the model on the asset file wasn't being updated, even when I removed the old one and added a new model the app was giving the output similar with the old model, e.g:
1. Add the model in asset dir
2. Build/run the app
3. No good results,
4. Deleted the old model, copy pasted again the new model with different name and still was not giving accurate results.
To me it looked that the load method was loading still the old model, I tried the clear the data of the app and then the model was "updated". I still can't reproduce the issue :/
